projection/src/main/scala/akka/projection/r2dbc/internal/R2dbcOffsetStore.scala

```diff
@@ -452,81 +453,89 @@ private[projection] class R2dbcOffsetStore(
+      .foldLeft((getInflight(), Vector.empty[Envelope])) { case ((inflight, filteredEnvelopes), envelope) =>
+        createRecordWithOffset(envelope) match {
+          case Some(recordWithOffset) =>
+            if (isAccepted(recordWithOffset, inflight))
+              (
+                inflight.updated(recordWithOffset.record.pid, recordWithOffset.record.seqNr),
+                filteredEnvelopes :+ envelope)
+            else
+              (inflight, filteredEnvelopes)
+          case None =>
+            (inflight, filteredEnvelopes :+ envelope)
+        }
       }
       ._2
   }
 
-  def isAccepted[Envelope](envelope: Envelope): Boolean =
-    isAccepted(envelope, getInflight())
-
-  private def isAccepted[Envelope](envelope: Envelope, currentInflight: Map[Pid, SeqNr]): Boolean = {
-    envelope match {
-      case eventEnvelope: EventEnvelope[_] if eventEnvelope.offset.isInstanceOf[TimestampOffset] =>
-        val pid = eventEnvelope.persistenceId
-        val seqNr = eventEnvelope.sequenceNr
-        val currentState = getState()
-        val timestampOffset = eventEnvelope.offset.asInstanceOf[TimestampOffset]
+  def isAccepted[Envelope](envelope: Envelope): Boolean = {
+    createRecordWithOffset(envelope) match {
+      case Some(recordWithOffset) => isAccepted(recordWithOffset, getInflight())
+      case None                   => true
+    }
+  }
 
-        val duplicate =
-          isDuplicate(
-            R2dbcOffsetStore
-              .Record(eventEnvelope.persistenceId, eventEnvelope.sequenceNr, timestampOffset.timestamp))
+  private def isAccepted[Envelope](recordWithOffset: RecordWithOffset, currentInflight: Map[Pid, SeqNr]): Boolean = {
+    val pid = recordWithOffset.record.pid
+    val seqNr = recordWithOffset.record.seqNr
+    val currentState = getState()
+    val timestampOffset = recordWithOffset.offset
+
+    val duplicate = isDuplicate(recordWithOffset.record)
+
+    if (duplicate) {
+      logger.debug("Filtering out duplicate: {}", recordWithOffset) // FIXME change to trace
+      false
+    } else if (recordWithOffset.strictSeqNr) {
+      // strictSeqNr == true is for event sourced
+      val ok =
+        (seqNr == 1L && (!currentState.byPid.contains(pid)) && (!currentInflight.contains(pid))) ||
+        JDuration
+          .between(timestampOffset.timestamp, timestampOffset.readTimestamp)
+          .compareTo(settings.acceptNewSequenceNumberAfterAge) >= 0 ||
+        seqNr == currentInflight.getOrElse(pid, Long.MinValue) + 1 ||
+        seqNr == currentState.byPid.get(pid).map(_.seqNr).getOrElse(Long.MinValue) + 1
+
+      if (!ok)
+        logger.debug("Filtering out rejected sequence number (might be accepted later): {}", recordWithOffset)
+
+      ok
+    } else {
+      // strictSeqNr == false is for durable state where each revision might not be visible
+      val ok =
+        seqNr > currentInflight.getOrElse(pid, 0L) &&
+        seqNr > currentState.byPid.get(pid).map(_.seqNr).getOrElse(0L)
```

Comment:
For durable state changes we accept anything that has a greater revision than previously known revision.